src/dynamic_batch_scheduler.cc

```diff
@@ -598,29 +598,34 @@ DynamicBatchScheduler::DelegateResponse(
   std::lock_guard<std::mutex> lock(completion_queue_mtx_);
   completion_queue_.emplace_back();
   auto queue_slot = &completion_queue_.back();
-  uint64_t request_hash = request->CacheKey();
   // Pass raw ptr to lambda for tracking stats from cache and updating
   // metric reporter on cache miss stats after insertion
   InferenceRequest* raw_request_ptr = request.get();
 
   request->SetResponseDelegator(
-      [this, queue_slot, request_hash, raw_request_ptr](
+      [this, queue_slot, raw_request_ptr](
           std::unique_ptr<InferenceResponse>&& response, const uint32_t flags) {
-        if (response_cache_enabled_) {
+        if (response_cache_enabled_ && raw_request_ptr->CacheKeyIsSet()) {
           // Cache insertion happens here because we need the backend to have
           // computed the inference response first in the case of cache miss
           auto cache = model_->Server()->GetResponseCache();
-          auto status = cache->Insert(request_hash, *response, raw_request_ptr);
-          if (status.IsOk()) {
+          auto status = cache->Insert(*response, raw_request_ptr);
+          // Cache Hit
+          if (status.StatusCode() == Status::Code::ALREADY_EXISTS) {
```

Comment:
NIT: I took a look at what the [compiler outputs](https://godbolt.org/z/adrGE9hv1) for this type of code and it apparently doesn't make much of a difference except when you define a local boolean before you test it (test_3 in the link). The compiler is able to look ahead and jump past the if-statement if the first expression of the compound conditional returns false. Nice to know but when turning on optimizations, [it will probably get reduced to that anyway](https://godbolt.org/z/4ab7hbx8b).

All this to say, I would suggest something like the following for readability.
```
// Failure to insert
bool failed_to_insert_request = status.StatusCode() != Status::Code::ALREADY_EXISTS && !status.IsOk();
if (failed_to_insert_request) {
    LOG_ERROR << "Failed to insert request_hash ["
                        << raw_request_ptr->CacheKey()
                        << "] into response cache: " << status.Message();
} // Do nothing as we already update cache hit statistics on Lookup
```